### ISSUE TITLE
init-declarations doesn't play nice with no-magic-numbers

### ISSUE THREAD

**Issue by jpickwell** [eslint/eslint#9064]
**Tell us about your environment**

* **ESLint Version: 4.3.0**
* **Node Version: 8.2.1**
* **yarn Version: 0.27.5**

**What parser (default, Babel-ESLint, etc.) are you using?**
Babel-ESLint

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
module.exports = {
  extends: 'eslint:recommended',
  parser: 'babel-eslint',
  plugins: [
    'babel',
    'html',
    'lodash',
    'you-dont-need-lodash-underscore'
  ],
  parserOptions: {
    ecmaVersion: 6,
    sourceType: 'module'
  },
  env: {
    browser: true,
    commonjs: true,
    es6: true,
    jquery: true,
    node: true
  },
  rules: {
    // excluding irrelevant rules
    'init-declarations': [2, 'never'],
    'no-console': 0,
    'no-magic-numbers': [2, { ignore: [-1, 0, 1] }],
    'no-var': 2
  }
}
```

</details>

**What did you do? Please include the actual source code causing the issue.**

```js
/* eslint-disable no-var */

// This is enough to reproduce.
var DWT_DIALOG_WIDTH = 392
```

**What did you expect to happen?**
No errors.

**What actually happened? Please include the actual, raw output from ESLint.**
```
  4:5  error  Variable 'DWT_DIALOG_WIDTH' should not be initialized on declaration  init-declarations

✖ 1 problem (1 error, 0 warnings)
```

If I separate the `var` from the assignment, then `no-magic-numbers` complains.

Note: The script that's exhibiting this issue is not being transpiled by Babel, so I want to avoid ES6 features (hence the `no-var` disable comment).

**Comment by not-an-aardvark** [comment]
Thanks for the issue. It does seem like there's a conflict here -- do you have any suggestions for how to resolve it?

**Comment by jpickwell** [comment]
An option could be added to `init-declarations`, say `ignorePattern` (default: `null`) where you can specify a RegExp which when matched the rule doesn't flag the assignment (e.g., `/^[0-9A-Z_]+$/`).

Alternatively, the option could be `includeMagicConstants` (default: `true`). This option would make the assumption that "magic" constants are named in all-caps with underscores (`[0-9A-Z_]+`). When the option is false, variables that match that regex would not be flagged by `init-declarations`.

Yet another path that could be taken is to add an option to `no-magic-numbers`, say `allowSeparateInitialization` (default: `false`). When `true`, this option would allow the declaration and initialization of "magic" constants to be separate statements.

I personally like the second option, but I can see users potentially wanting to exclude other variables with the use of the RegExp option. I threw in the third option for good measure, but I personally would never want to separate the declaration and initialization of "magic" constants.

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get consensus from the team, so I'm closing it. We define [consensus](http://eslint.org/docs/maintainer-guide/issues.html#consensus) as having three :+1:s from team members, as well as a team member willing to champion the proposal. This is a high bar by design -- we can't realistically accept and maintain every feature request in the long term, so we only accept feature requests which are useful enough that there is consensus among the team that they're worth adding.

Since ESLint is pluggable and can load custom rules at runtime, the lack of consensus among the ESLint team doesn't need to be a blocker for you using this in your project, if you'd find it useful. It just means that you would need to [implement the rule](http://eslint.org/docs/developer-guide/working-with-rules) yourself, rather than using a bundled rule that is packaged with ESLint.
